crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -165,19 +189,14 @@ where
 
     fn headers_range(&self, range: impl RangeBounds<BlockNumber>) -> ProviderResult<Vec<Header>> {
         let mut headers = Vec::new();
-        let start = match range.start_bound() {
-            Bound::Included(&n) => n,
-            Bound::Excluded(&n) => n + 1,
-            Bound::Unbounded => 0,
-        };
-        let end = match range.end_bound() {
-            Bound::Included(&n) => n,
-            Bound::Excluded(&n) => n - 1,
-            Bound::Unbounded => self.canonical_in_memory_state.get_canonical_block_number(),
-        };
+        let (start, end) = self.convert_range_bounds(range, || {
+            self.canonical_in_memory_state.get_canonical_block_number()
+        });
 
         for num in start..=end {
             if let Some(block_state) = self.canonical_in_memory_state.state_by_number(num) {
+                // TODO: there might be an update between loop iterations, we
+                // need to handle that situation.
```

Comment:
it might be that we need to move this feature into the canonicalmemory type directly.

will track in a new issue